From e432c77f2361aca5df99f3f7dcae0d1ae94a948a Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 15 Dec 2022 06:03:01 +0000
Subject: [PATCH 1174/1398] [Huawei] Revert "[Huawei] mm/sharepool: check
 permission of task to operate spg_node->proc_node"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I650K6

--------------------------------

This reverts commit 63e393693738ba5fbf33b0694d6f6dedb27c7adc.
Reasons to revert this patch:
1. This patch introduced a suspected problem that
   sp_group_node_permission_check may meet a NULL task->mm when the task
   is exiting. Actually this would not happen becasue we would hold
   sp_group_sem in sp_group_exit before we set task->mm to NULL and
   sp_group_node_permission_check is called with sp_group_sem hold.
   This is hard to understand.
2. This patch check the unsupported scene at the point we operate on
   spg_node->proc_node. Actually we could do that just at the point we
   get the task's mm_struct. This would much more clear and could avoid
   changing a series of inline function's parameter.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 67 +++++++++----------------------------------------
 1 file changed, 12 insertions(+), 55 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index fc0877256559..bd1ef2cc8803 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -237,7 +237,6 @@ struct sp_group_master {
 	/* list head of sp_node */
 	struct list_head node_list;
 	struct mm_struct *mm;
-	pid_t tgid;
 	/*
 	 * Used to apply for the shared pool memory of the current process.
 	 * For example, sp_alloc non-share memory or k2task.
@@ -495,9 +494,9 @@ static struct sp_mapping *sp_mapping_find(struct sp_group *spg,
 static struct sp_group *create_spg(int spg_id, unsigned long flag);
 static void free_new_spg_id(bool new, int spg_id);
 static void free_sp_group_locked(struct sp_group *spg);
-static struct sp_group_node *group_add_task(struct task_struct *task, struct sp_group *spg,
+static struct sp_group_node *group_add_task(struct mm_struct *mm, struct sp_group *spg,
 					    unsigned long prot);
-static int init_local_group(struct task_struct *task, struct mm_struct *mm)
+static int init_local_group(struct mm_struct *mm)
 {
 	int spg_id, ret;
 	struct sp_group *spg;
@@ -528,7 +527,7 @@ static int init_local_group(struct task_struct *task, struct mm_struct *mm)
 	sp_mapping_attach(master->local, sp_mapping_normal);
 	sp_mapping_attach(master->local, sp_mapping_ro);
 
-	spg_node = group_add_task(task, spg, PROT_READ | PROT_WRITE);
+	spg_node = group_add_task(mm, spg, PROT_READ | PROT_WRITE);
 	if (IS_ERR(spg_node)) {
 		/* The spm would be released while destroying the spg */
 		ret = PTR_ERR(spg_node);
@@ -561,12 +560,11 @@ static int sp_init_group_master_locked(struct task_struct *tsk, struct mm_struct
 	INIT_LIST_HEAD(&master->node_list);
 	master->count = 0;
 	master->mm = mm;
-	master->tgid = tsk->tgid;
 	sp_init_group_master_stat(tsk->tgid, mm, &master->instat);
 	mm->sp_group_master = master;
 	sp_add_group_master(master);
 
-	ret = init_local_group(tsk, mm);
+	ret = init_local_group(mm);
 	if (ret)
 		goto free_master;
 
@@ -1341,34 +1339,13 @@ static int insert_spg_node(struct sp_group *spg, struct sp_group_node *node)
 	return 0;
 }
 
-static void __delete_spg_node(struct sp_group *spg, struct sp_group_node *node)
+/* the caller must down_write(&spg->rw_lock) */
+static void delete_spg_node(struct sp_group *spg, struct sp_group_node *node)
 {
 	list_del(&node->proc_node);
 	spg->proc_num--;
 }
 
-/*
- * if process A clone B with CLONE_VM and *NO* CLONE_THREAD, and not do exec().
- * B will have a different tgid than A, but share mm with A.
- * In this case, we need check task->tgid is matched with sp_group_master->tgid.
- * Any operation on spg_node->proc_list needs to pass the permission check.
- */
-static int sp_group_node_permission_check(struct task_struct *task)
-{
-	return task->tgid == task->mm->sp_group_master->tgid;
-}
-
-/* the caller must down_write(&spg->rw_lock) */
-static void delete_spg_node(struct task_struct *task, struct sp_group_node *node)
-{
-	if (!sp_group_node_permission_check(task)) {
-		pr_info("delete task from group failed, tgid is not matched\n");
-		return;
-	}
-
-	__delete_spg_node(node->spg, node);
-}
-
 /* the caller must hold sp_group_sem */
 static void free_spg_node(struct mm_struct *mm, struct sp_group *spg,
 	struct sp_group_node *spg_node)
@@ -1381,7 +1358,8 @@ static void free_spg_node(struct mm_struct *mm, struct sp_group *spg,
 	kfree(spg_node);
 }
 
-static struct sp_group_node *__group_add_task(struct mm_struct *mm, struct sp_group *spg,
+/* the caller must hold sp_group_sem and down_write(&spg->rw_lock) in order */
+static struct sp_group_node *group_add_task(struct mm_struct *mm, struct sp_group *spg,
 					    unsigned long prot)
 {
 	struct sp_group_node *node;
@@ -1400,20 +1378,6 @@ static struct sp_group_node *__group_add_task(struct mm_struct *mm, struct sp_gr
 	return node;
 }
 
-/* the caller must hold sp_group_sem and down_write(&spg->rw_lock) in order */
-static struct sp_group_node *group_add_task(struct task_struct *task, struct sp_group *spg,
-					    unsigned long prot)
-{
-	struct mm_struct *mm = task->mm;
-
-	if (!sp_group_node_permission_check(task)) {
-		pr_info("add task into group failed, tgid is not matched\n");
-		return ERR_PTR(-EPERM);
-	}
-
-	return __group_add_task(mm, spg, prot);
-}
-
 /**
  * mg_sp_group_add_task() - Add a process to an share group (sp_group).
  * @tgid: the tgid of the task to be added.
@@ -1545,7 +1509,7 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		goto out_drop_group;
 	}
 
-	node = group_add_task(tsk, spg, prot);
+	node = group_add_task(mm, spg, prot);
 	if (unlikely(IS_ERR(node))) {
 		up_write(&spg->rw_lock);
 		ret = PTR_ERR(node);
@@ -1629,7 +1593,7 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 	spin_unlock(&sp_area_lock);
 
 	if (unlikely(ret))
-		delete_spg_node(tsk, node);
+		delete_spg_node(spg, node);
 	up_write(&spg->rw_lock);
 
 	if (unlikely(ret))
@@ -1731,7 +1695,7 @@ int mg_sp_group_del_task(int tgid, int spg_id)
 	down_write(&spg->rw_lock);
 	if (list_is_singular(&spg->procs))
 		is_alive = spg->is_alive = false;
-	delete_spg_node(tsk, spg_node);
+	delete_spg_node(spg, spg_node);
 	sp_group_drop(spg);
 	up_write(&spg->rw_lock);
 	if (!is_alive)
@@ -4352,12 +4316,6 @@ int sp_group_exit(void)
 		return 0;
 	}
 
-	if (!sp_group_node_permission_check(current)) {
-		up_write(&sp_group_sem);
-		pr_info("tgid is not matched in exit\n");
-		return 1;
-	}
-
 	list_for_each_entry_safe(spg_node, tmp, &master->node_list, group_node) {
 		spg = spg_node->spg;
 
@@ -4365,8 +4323,7 @@ int sp_group_exit(void)
 		/* a dead group should NOT be reactive again */
 		if (spg_valid(spg) && list_is_singular(&spg->procs))
 			is_alive = spg->is_alive = false;
-		/* the permission is checked */
-		__delete_spg_node(spg, spg_node);
+		delete_spg_node(spg, spg_node);
 		up_write(&spg->rw_lock);
 
 		if (!is_alive)
-- 
2.17.1

